--- ./src/edu/yale/its/tp/cas/client/filter/CASFilter.java.orig	2005-10-21 15:12:01.000000000 +0200
+++ ./src/edu/yale/its/tp/cas/client/filter/CASFilter.java	2005-10-21 10:41:08.000000000 +0200
@@ -94,6 +94,12 @@
      */
     public final static String LOGIN_INIT_PARAM = "edu.yale.its.tp.cas.client.filter.loginUrl";
     
+    /** The name of the filter initialization parameter the value of which should be the https: address
+     * of the CAS Logout servlet.  Optional parameter, but required for successful redirection of 
+     * logout requests
+     */
+    public final static String LOGOUT_INIT_PARAM = "edu.yale.its.tp.cas.client.filter.logoutUrl";
+    
     /** The name of the filter initialization parameter the value of which must be the https: address
      * of the CAS Validate servlet.  Must be a CAS 2.0 validate servlet (CAS 1.0 non-XML won't suffice).
      * Required parameter.
@@ -165,6 +171,8 @@
 
     /** Secure URL whereat CAS offers its login service. */
     private String casLogin;
+    /** Secure URL whereat CAS offers its logout service. */
+    private String casLogout;
     /** Secure URL whereat CAS offers its CAS 2.0 validate service */
     private String casValidate;
     /** Filtered service URL for use as service parameter to login and validate */
@@ -196,6 +204,9 @@
         casLogin =
             config.getInitParameter(
                 LOGIN_INIT_PARAM);
+        casLogout =
+            config.getInitParameter(
+                LOGOUT_INIT_PARAM);
         casValidate =
             config.getInitParameter(
                 VALIDATE_INIT_PARAM);
@@ -302,6 +313,19 @@
             return;
         }
 
+	// Is this a logout request ? If so, redirect to the logout URL
+	if (request.getParameter("ssoLogout") != null) {
+	    log.trace("logout from CAS SSO requested");
+	    // remove CAS Filter attributes
+	    HttpSession session = ((HttpServletRequest) request).getSession();
+	    session.removeAttribute(CAS_FILTER_USER);
+	    session.removeAttribute(CAS_FILTER_RECEIPT);
+	    // then redirect to CAS logout URL
+	    ((HttpServletResponse) response).sendRedirect(casLogout);
+	    // abort chain
+	    return;
+	}
+
         // Wrap the request if desired
         if (wrapRequest) {
         		log.trace("Wrapping request with CASFilterRequestWrapper.");
@@ -538,6 +562,12 @@
             sb.append(" casLogin=NULL!!!!!");
         }
         
+        if (this.casLogout != null) {
+            sb.append(" casLogout=[");
+            sb.append(this.casLogout);
+            sb.append("]");
+        } 
+
         if (this.casProxyCallbackUrl != null) {
             sb.append(" casProxyCallbackUrl=[");
             sb.append(casProxyCallbackUrl);
